File: packages/pdf-gen-py/pdf_gen/templates/certificate.py
from typing import Any, Dict

from ..config import PDFConfig, Theme
from ..surface import Surface
from ..utils.validate import require_fields, format_date


def render_certificate(surface: Surface, config: PDFConfig, data: Dict[str, Any], theme: Theme) -> None:
    require_fields(data, ['title', 'recipient_name', 'achievement_text', 'issuer_name'], 'certificate')

    page_width = surface.page_width
    page_height = surface.page_height
    center_x = page_width / 2
    border_color = data.get('border_color', theme.primary)

    if data.get('decorative_elements', True):
        from reportlab.lib.colors import HexColor

        surface.canvas.saveState()
        surface.canvas.setStrokeColor(HexColor(border_color))
        surface.canvas.setLineWidth(3)
        surface.canvas.rect(30, 30, page_width - 60, page_height - 60, fill=0, stroke=1)

        surface.canvas.setStrokeColor(HexColor(theme.accent))
        surface.canvas.setLineWidth(1)
        surface.canvas.rect(42, 42, page_width - 84, page_height - 84, fill=0, stroke=1)
        surface.canvas.restoreState()

        for cx, cy_top in [(42, 42), (page_width - 42, 42), (42, page_height - 42), (page_width - 42, page_height - 42)]:
            surface.circle(cx, cy_top, 4, fill=theme.accent)

    y = page_height * 0.14

    y = surface.text(60, y, data['title'], font='Helvetica-Bold', size=30, color=border_color, align='center', width=page_width - 120)
    y += 20

    y = surface.text(60, y, 'This certificate is proudly presented to', font='Helvetica', size=15, color=theme.text, align='center', width=page_width - 120)
    y += 20

    y = surface.text(60, y, data['recipient_name'], font='Helvetica-Bold', size=30, color=theme.primary_dark, align='center', width=page_width - 120)
    y += 6

    surface.line(center_x - 120, y, center_x + 120, y, color=theme.accent, width=1.5)
    y += 24

    surface.text(100, y, data['achievement_text'], font='Helvetica', size=13, color=theme.text, align='center', width=page_width - 200)

    bottom_y = page_height - 150

    surface.text(100, bottom_y, f"Date: {format_date(data.get('issue_date'))}", font='Helvetica', size=11, color=theme.muted, width=220)
    cert_number = data.get('certification_number')
    if cert_number:
        surface.text(100, bottom_y + 16, f'Certificate No: {cert_number}', font='Helvetica', size=11, color=theme.muted, width=220)

    surface.line(100, bottom_y - 10, 320, bottom_y - 10, color=theme.border)
    surface.line(page_width - 320, bottom_y - 10, page_width - 100, bottom_y - 10, color=theme.border)

    surface.text(page_width - 320, bottom_y, data['issuer_name'], font='Helvetica-Bold', size=11, color=theme.text, align='right', width=220)
    surface.text(page_width - 320, bottom_y + 16, 'Authorized Signature', font='Helvetica', size=9, color=theme.muted, align='right', width=220)

File: packages/pdf-gen-js/src/templates/certificate.ts
import PDFDocument from 'pdfkit';
import { PDFConfig } from '../config';

export interface CertificateData {
  title: string;
  recipientName: string;
  achievementText: string;
  issuerName: string;
  issueDate: Date;
  certificationNumber?: string;
  borderColor?: string;
  decorativeElements?: boolean;
}

export function renderCertificate(pdf: PDFDocument, config: PDFConfig, data: CertificateData): void {
  const pageWidth = pdf.page.width;
  const pageHeight = pdf.page.height;
  const centerX = pageWidth / 2;
  const borderColor = data.borderColor || '#1a5f7a';

  // Decorative border
  if (data.decorativeElements !== false) {
    pdf.strokeColor(borderColor).lineWidth(2);
    pdf.rect(40, 40, pageWidth - 80, pageHeight - 80).stroke();

    pdf.lineWidth(1);
    pdf.rect(60, 60, pageWidth - 120, pageHeight - 120).stroke();
  }

  // Title
  pdf.font('Helvetica-Bold', 32)
    .fillColor(borderColor)
    .text(data.title, 0, 100, { align: 'center', width: pageWidth });

  // This certifies
  pdf.fillColor('#000000').font('Helvetica', 16).text('This is to certify that', 0, 180, {
    align: 'center',
    width: pageWidth
  });

  // Recipient name
  pdf.font('Helvetica-Bold', 28)
    .fillColor(borderColor)
    .text(data.recipientName, 0, 230, { align: 'center', width: pageWidth });

  // Achievement text
  pdf.fillColor('#000000')
    .font('Helvetica', 13)
    .text(data.achievementText, 100, 300, { align: 'center', width: pageWidth - 200 });

  // Date and issuer
  const bottomSectionY = pageHeight - 180;
  pdf.font('Helvetica', 11);
  pdf.text(`Date: ${data.issueDate.toLocaleDateString()}`, 100, bottomSectionY);

  if (data.certificationNumber) {
    pdf.text(`Certificate #: ${data.certificationNumber}`, 100, bottomSectionY + 30);
  }

  // Signature line
  pdf.moveTo(100, bottomSectionY + 80).lineTo(250, bottomSectionY + 80).stroke();
  pdf.font('Helvetica', 10).text('Authorized Signature', 100, bottomSectionY + 85);

  // Issuer
  pdf.moveTo(pageWidth - 250, bottomSectionY + 80).lineTo(pageWidth - 100, bottomSectionY + 80).stroke();
  pdf.text(data.issuerName, pageWidth - 250, bottomSectionY + 85, { width: 150 });
}
